File: exclusive-exec-prototype-example/src/main/java/io/averkhoglyad/proto/example/SomeComponent.java
package io.averkhoglyad.proto.example;

import io.averkhoglyad.proto.exclusive.Exclusive;

public class SomeComponent {

    @Exclusive
    public void test0() {
        waitAndThen(() -> System.out.println("test0()"));
    }

    @Exclusive
    public void test1(Object arg1) {
        waitAndThen(() -> System.out.println("test1(" + arg1 + ")"));
    }

    @Exclusive
    public void test2(Object arg1, Object  arg2) {
        waitAndThen(() -> System.out.println("test2(" + arg1 + ", " + arg2 + ")"));
    }

    @Exclusive
    public void test1_2(@Exclusive.Key Object arg1, Object arg2) {
        waitAndThen(() -> System.out.println("test1_2(" + arg1 + ", " + arg2 + ")"));
    }

    protected void waitAndThen(Runnable fn) {
        try {
            Thread.sleep(3000);
            fn.run();
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
    }
}

File: src/main/kotlin/io/averkhoglyad/proto/SomeComponent.kt
package io.averkhoglyad.proto

import io.averkhoglyad.proto.exclusive.Exclusive

@Exclusive
open class SomeComponent {
    open fun test0() {
        waitAndThen { println("test0()") }
    }

    open fun test1(arg1: Any) {
        waitAndThen { println("test1($arg1)") }
    }

    open fun test2(arg1: Any, arg2: Any) {
        waitAndThen { println("test2($arg1, $arg2)") }
    }

    open fun test1_2(@Exclusive.Key arg1: Any, arg2: Any) {
        waitAndThen { println("test1_2($arg1, $arg2)") }
    }

    private fun waitAndThen(fn: Runnable) {
        try {
            Thread.sleep(3000)
            fn.run()
        } catch (e: InterruptedException) {
            throw RuntimeException(e)
        }
    }
}
